### PR TITLE
fix: gf over a 'file://...' path and #3310 issue

### LINKED ISSUE #3310: gf: 'try to find it with the same extension'-code doesn't work
**Describe the bug**

`gf` under the `require('some_file')` in .js file says "Vim: This file /Users/vvp/test/some_file.js does not exist.".

**To Reproduce**
1.  `echo '// some_file' > file.js && touch some_file.js`
2.  `code file.js`
3.  `gf` under the `some_file`
4.  See error

**Expected behavior**
some_file.js has opened

**Environment (please complete the following information):**
- Extension (VsCodeVim) version: 0.17.3
- VSCode version: 1.30.1
- OS: Mac OS 10.13.4

### PR DESCRIPTION
**Pull request by pikulev**
**What this PR does / why we need it**:
A bunch of solutions:
- #3310 issue 
- now `gf` works under a path to a file that has no extension AND with a dot in the middle of his name, i.e. under the `some.file` `gf` will open `some.file.${same extension}`
- let `gf` works under a 'file://..' path

**Which issue(s) this PR fixes**
 fixes #3310

### CODE REVIEW

**Comment by jpoon** [comment]
Fixes https://github.com/VSCodeVim/Vim/issues/3233

File: src/cmd_line/commands/file.ts [hunk 1]
```diff
@@ -84,26 +84,27 @@ export class FileCommand extends node.CommandBase {
...
+          : <string>untildify(this._arguments.name);
       filePath = path.isAbsolute(this._arguments.name)
         ? this._arguments.name
         : path.join(path.dirname(editorFilePath), this._arguments.name);
 
       if (filePath !== editorFilePath && !(await util.promisify(fs.exists)(filePath))) {
-        // if file does not exist and does not have an extension
-        // try to find it with the same extension
-        if (path.extname(filePath) === '') {
+        if (this._arguments.createFileIfNotExists) {
```
Comment (jpoon) [inline src/cmd_line/commands/file.ts]: I don't yet see the root cause of the bug. Can you walk through the logic for the fix here?

Comment (jpoon) [reply src/cmd_line/commands/file.ts]: In ts/js, it's typical not to provide the file extension of the file on imports. For instance,

```
import { configuration } from '../../src/configuration/configuration';
```

on a `gf`, filePath should equal something like ` ../../src/configuration/configuration`. The expected behaviour is that we should:
1) try to find the file as is (configuration)
2) if it doesn't have an extension, try to find it with the same extension as the current file (configuration.ts)
3) then if `createFileIfNotExists=true`, we create the file.

WIth your proposed changed, we would always be creating a new file (presuming `createFileIfNotExists=true`). IOW, I think we need another check following the if block of 94 to see if we found it with the same extension.

So maybe something like this:

```
      // Using a filename, open or create the file
      this._arguments.name = <string>untildify(this._arguments.name);
      filePath = path.isAbsolute(this._arguments.name)
        ? this._arguments.name
        : path.join(path.dirname(editorFilePath), this._arguments.name);

      isFound = false;
      if (filePath !== editorFilePath) {
        isFound |= await util.promisify(fs.exists)(filePath)

        if (!isFound) {
            // if file does not exist and does not have an extension
            // try to find it with the same extension
            if (path.extname(filePath) === '') {
              const pathWithExt = filePath + path.extname(editorFilePath);
              isFound |= await util.promisify(fs.exists)(pathWithExt)

              if (isFound) {
                  filePath = pathWithExt;
              }
          }
        }

        // if not found, and createFileIfNotExists = true, create file
        // else if not found, error
}
```

Comment (pikulev) [reply src/cmd_line/commands/file.ts]: Thank you! You are right. Actually, my thoughts about a role of the createFileIfNotExists flag were slightly different.  I've just recovered the logic you are described.

PS: my congratulations :) best wishes for you

File: src/cmd_line/commands/file.ts [hunk 2]
```diff
@@ -83,21 +83,22 @@ export class FileCommand extends node.CommandBase {
...
-        // try to find it with the same extension
-        if (path.extname(filePath) === '') {
-          const pathWithExt = filePath + path.extname(editorFilePath);
-          if (await util.promisify(fs.exists)(pathWithExt)) {
-            filePath = pathWithExt;
-          }
-        }
-        if (this._arguments.createFileIfNotExists) {
+        // if file does not exist try to find it with the same extension
+        const pathWithExt = filePath + path.extname(editorFilePath);
```
[the remark this answers is not in the document]
Comment (pikulev) [reply src/cmd_line/commands/file.ts]: Oh, I got it. You are about

> _if it doesn't have an extension_, try to find it with the same extension as the current file (configuration.ts)

The problem was here:
```js
        // if file does not exist and does not have an extension
        // try to find it with the same extension
        if (path.extname(filePath) === '') { ...
```

The extname method returns simply all chars after dot. It is not correct in cases when filename has a dot in the middle. I.e. `import Constants from 'constants.something'` in js file will be use 'something' as a 'same-ext'

So. Now the logic is as follows

1. try to find the file as is (configuration)
2. if 'configuration' does not exist try to find it with the same extension as the current file (configuration.ts)
3. then if createFileIfNotExists=true, we create the file.

[the remark this answers is not in the document]
Comment (jpoon) [reply src/cmd_line/commands/file.ts]: > it is not correct in the cases when filename has a dot i the middle

Not according to the docs https://nodejs.org/api/path.html

[the remark this answers is not in the document]
Comment (pikulev) [reply src/cmd_line/commands/file.ts]: > in its current state, you do this irrespective of whether the initial file has an extension
 
Agree. By that we have extra check `await util.promisify(fs.exists)(pathWithExt)` whether the pathWithExt empty or not.

> we should really write UTs for this
Absolutely. There is still a lot of small problems. I. e. fs.exists checks dirs too, so if you are trying go to file than does not exist and you have a directory with the same name you doesn't get an error and nothing happened.

[the remark this answers is not in the document]
Comment (jpoon) [reply src/cmd_line/commands/file.ts]: > Hm. Can you please check out https://repl.it/repls/HeavyLiquidLifecycles <- that is what I mean.

```
const path = require('path');
console.log(path.extname('./config'));
```

that behaviour is inline with the docs (https://nodejs.org/api/path.html#path_path_extname_path) in that it returns an empty string.

**Comment by jpoon** [comment]
Closing as https://github.com/VSCodeVim/Vim/pull/3396 supersedes this.

File: src/cmd_line/commands/file.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jpoon) [reply src/cmd_line/commands/file.ts]: Ah good point, I'll fix that in my next PR

**Pull request merged**